### PR TITLE
Implement WSS dialing

### PR DESCRIPTION
**Pull request by albrow**
See #70. This PR adds support for dialing (but not listening on) WSS addresses. Listening can be accomplished via a proxy such as nginx, so I would rather not implement it here for the sake of time. I borrowed some code from https://github.com/libp2p/go-ws-transport/pull/46 (thank you @Jorropo!), addressed some of the concerns from that PR, and added a test which actually dials one of the public bootstrap nodes using WSS.

I'm sure that I am missing some things and there are a few smaller issues to sort out. The basics are working so I decided to go ahead and open a PR.

### CODE REVIEW

File: addrs.go [hunk 1]
```diff
@@ -45,17 +55,29 @@ func ParseWebsocketNetAddr(a net.Addr) (ma.Multiaddr, error) {
 		return nil, fmt.Errorf("not a websocket address")
 	}
 
-	tcpaddr, err := net.ResolveTCPAddr("tcp", wsa.Host)
-	if err != nil {
-		return nil, err
+	// Detect if host is IP address or DNS
+	var tcpma ma.Multiaddr
+	if ip := net.ParseIP(wsa.Hostname()); ip != nil {
```
Comment (albrow) [inline addrs.go]: I had to change this code because we cannot lookup DNS addresses in the browser. `ResolveTCPAddr` calls `lookupIP` which is [not implemented in JavaScript/Wasm environments](https://github.com/golang/go/blob/e948d2b73ede67f12bff9e4d050f0e1425163010/src/net/lookup_fake.go#L22). This code seems to work but I still have two questions:

1. Does libp2p _need_ to resolve the DNS name here for some reason?
2. Suppose `a` contains a domain name instead of an IP address. How can I differentiate between dns4 and dns6?

File: websocket_test.go [hunk 2]
```diff
@@ -22,25 +23,48 @@ func TestCanDial(t *testing.T) {
...
+	if !d.CanDial(addrWss) {
+		t.Fatal("expected to match secure websocket maddr, but did not")
+	}
+	if d.CanDial(addrTCP) {
 		t.Fatal("expected to not match tcp maddr, but did")
 	}
 }
 
+func TestDialWss(t *testing.T) {
+	raddr, err := ma.NewMultiaddr("/dns4/nyc-1.bootstrap.libp2p.io/tcp/443/wss")
```
Comment (albrow) [inline websocket_test.go]: Is it okay to leave this in the tests? Usually I would say it is not proper to dial out to an independent server in this kind of test, but it did help me find some bugs. Without this, we cannot easily test that dialing actually works.

File: conn_native.go [hunk 3]
```diff
@@ -104,11 +104,13 @@ func (c *Conn) Close() error {
 }
 
 func (c *Conn) LocalAddr() net.Addr {
-	return NewAddr(c.Conn.LocalAddr().String())
+	// TODO(albrow): Detect if WSS or WS
+	return NewAddr("ws", c.Conn.LocalAddr().String())
```
Comment (albrow) [inline conn_native.go]: Is there a way to detect whether the underlying `gorilla/websocket` connection is using `ws` or `wss`?

File: websocket.go [hunk 4]
```diff
@@ -27,9 +36,21 @@ var WsCodec = &manet.NetCodec{
 	ConvertMultiaddr: ConvertWebsocketMultiaddrToNetAddr,
 	ParseNetAddr:     ParseWebsocketNetAddr,
 }
+var WssCodec = &manet.NetCodec{
+	NetAddrNetworks:  []string{"websocket secure"},
```
Comment (albrow) [inline websocket.go]: Is `websocket secure` okay for the network name? The other PR was using `websocket+tls`.

File: websocket_native.go [hunk 5]
```diff
@@ -22,13 +23,24 @@ var upgrader = ws.Upgrader{
...
 }
 
+var wsDialer *ws.Dialer
+
+func init() {
+	// Initialize our own wsDialer which skips TLS verification.
+	wsDialer = new(ws.Dialer)
+	(*wsDialer) = *ws.DefaultDialer
+	wsDialer.TLSClientConfig = &tls.Config{
+		InsecureSkipVerify: true,
```
Comment (albrow) [inline websocket_native.go]: I'm not 100% sure if this is the correct way to do things. Removing this line causes the tests to fail with the following error:

```
x509: certificate signed by unknown authority
```

There is also the related question of how WSS should work with secio. Hypothetically, secio is not required if we are using WSS, but we do need to consider how to verify that the peer we are connected to has the correct identity (i.e. that they have a private key corresponding to the peer ID).

Comment (albrow) [reply websocket_native.go]: Also AFAIK there is no equivalent of `InsecureSkipVerify` in the browser, as it would violate the security model. I added a test in this PR that shows that dialing a WSS multiaddr in the browser does work. I wonder if it will cause problems when hosted on an HTTPS domain? Or maybe the certificate verification used in browsers is different than what Go is doing?

**Comment by pgebheim** [comment]
The main issue with this PR is that any sort of listening requires certificates signed by a CA. So unless everyone is generating Let's Encrypt certs or something for their clients, this basically won't work.

So, @albrow -- that error will happen whenever a client isn't actually serving using a CA-signed cert. That will cause a browser to balk, then throw client errors. In the case of the Go code it looks like these security errors actually throw and terminate the 0x mesh.

I may be mistaken but this seems like a major issue -- as clients who are discovered via a bootnode tell peers of their multiaddrs, and if they're WS will be pro-actively dialed and then cause errors.

**Comment by albrow** [comment]
@pgebheim While listening on WSS would be a good feature to add to `go-ws-transport`, it's not strictly necessary since SSL can be handled via a proxy ([e.g. nginx](https://www.serverlab.ca/tutorials/linux/web-servers-linux/how-to-proxy-wss-websockets-with-nginx/)). Our plan is to get our own SSL certificate and use nginx as a proxy our bootstrap nodes and relay nodes, which will then be able to accept WSS connections. We will rely on these in the near term to establish connections and route traffic between any browser peers in the network. See https://docs.libp2p.io/concepts/circuit-relay/.

> In the case of the Go code it looks like these security errors actually throw and terminate the 0x mesh.

I believe Mesh only throws when it can't connect to the bootstrap nodes (since it wouldn't really be able to continue). We should be able to catch the error in any other circumstances and just log it. As an optimization we can potentially detect if we are being hosted on HTTPS and if so not even attempt to dial WS endpoints and only try dialing WSS.

**Comment by pgebheim** [comment]
Ok this makes sense re: bootstrap nodes.  It seems like in this case it errored even though there are other was bootstrap nodes, I can attempt to verify. But iirc there is a specific error for when no bootstrap nodes are connectable.

It would be nice to filter out non secure dials because or else clients are just doing a lot of work and producing errors.

File: addrs.go [hunk 6]
```diff
@@ -22,10 +22,11 @@ func (addr *Addr) Network() string {
 }
 
 // NewAddr creates a new Addr using the given host string
-func NewAddr(host string) *Addr {
+func NewAddr(scheme string, host string) *Addr {
```
Comment (raulk) [inline addrs.go]: I'd prefer to deprecate this function and add a new one that accepts the scheme and a flag indicating whether this is IPv4 or IPv6. Then we can address your TODO in line 73 cleanly.

File: addrs.go [hunk 7]
```diff
@@ -45,17 +55,29 @@ func ParseWebsocketNetAddr(a net.Addr) (ma.Multiaddr, error) {
...
+		if err != nil {
+			return nil, err
+		}
+		tcpma, err = manet.FromNetAddr(tcpaddr)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Assume DNS name
+		var err error
```
Comment (raulk) [inline addrs.go]: nit: Hoist err to the top, or even better, used named return values.

File: addrs.go [hunk 8]
```diff
@@ -64,10 +86,17 @@ func ParseWebsocketNetAddr(a net.Addr) (ma.Multiaddr, error) {
 }
 
 func parseMultiaddr(a ma.Multiaddr) (string, error) {
-	_, host, err := manet.DialArgs(a)
+	p := a.Protocols()
```
Comment (raulk) [inline addrs.go]: Could you explain why this change? Why not use `DialArgs` and then [`net.SplitHostPort`](https://golang.org/pkg/net/#SplitHostPort) to get the host and the port separately? You can then prepend `ws` or `wss` based on the presence of the protocol.

File: addrs.go [hunk 9]
```diff
@@ -45,17 +55,29 @@ func ParseWebsocketNetAddr(a net.Addr) (ma.Multiaddr, error) {
...
+		}
+		tcpma, err = manet.FromNetAddr(tcpaddr)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Assume DNS name
+		var err error
+		// TODO(albrow): What about dns6?
+		tcpma, err = ma.NewMultiaddr(fmt.Sprintf("/dns4/%s/tcp/%s", wsa.Hostname(), wsa.Port()))
```
Comment (raulk) [inline addrs.go]: A `/dns` address should work and resolve both A and AAAA records.

File: websocket.go (hunk 4 above)
Comment (raulk) [reply websocket.go]: WebSocket Secure is the official name, so it LGTM.

File: websocket.go [hunk 10]
```diff
@@ -16,9 +17,17 @@ import (
 //
 // Deprecated: use `ma.ProtocolWithCode(ma.P_WS)
 var WsProtocol = ma.ProtocolWithCode(ma.P_WS)
+var WssProtocol = ma.Protocol{
```
[the remark this answers is not in the document]
Comment (raulk) [reply websocket.go]: It wouldn't hurt!

File: websocket_native.go (hunk 5 above)
Comment (raulk) [reply websocket_native.go]: > There is also the related question of how WSS should work with secio.

For WSS dials and inbound connections, we wouldn't perform the security upgrade, only the multiplexing upgrade.

File: websocket_test.go (hunk 2 above)
Comment (raulk) [reply websocket_test.go]: It'll make this test fail in offline scenarios. I rather test whether the host is resolvable, and if not, `t.Skip()` the test. Do we have a ping-pong test? (libp2p WSS listener and dialer interacting with one another)

File: conn_native.go (hunk 3 above)
Comment (raulk) [reply conn_native.go]: You could add a field in `Conn`?

File: websocket_native.go (hunk 5 above)
Comment (raulk) [reply websocket_native.go]: > I'm not 100% sure if this is the correct way to do things. Removing this line causes the tests to fail with the following error:

I would prefer to stay secure by default, and enable `InsecureSkipVerify` if a flag is passed to the constructor of the transport (similar to what I described in #70). You can store this setting in the `WebSocketTransport` struct.

File: websocket_test.go (hunk 2 above)
Comment (albrow) [reply websocket_test.go]: Since I didn't implement listening in this PR, it is pretty hard to write a ping-pong test the way we normally would (e.g. with two hosts running locally). Even if we did have listening there would be some challenges because we would have to use a domain name with an SSL certificate.

My vote is to `t.Skip()` this test for now and manually enable it whenever we need to test dialing again in the future. What do you think?

Comment (albrow) [reply websocket_test.go]: @raulk I updated the tests so that if we can't resolve the host, we call `t.Skip`. I also improved the test by opening a stream to make sure that the connection is _really_ working.

File: addrs.go (hunk 8 above)
Comment (albrow) [reply addrs.go]: I just copied this over from #46.

`DialArgs` returns the following error: 

```
/dns/nyc-1.bootstrap.libp2p.io/tcp/443/wss is not a 'thin waist' address
```

Do you know how to address that?

File: addrs.go (hunk 6 above)
Comment (albrow) [reply addrs.go]: Done. We don't need to differentiate between IPv4 and IPv6 anymore if we can just use `/dns`. Not sure what to name this new function. Currently it's:

```go
// NewAddrWithScheme creates a new Addr using the given host string. isSecure
// should be true for WSS connections and false for WS.
func NewAddrWithScheme(host string, isSecure bool) *Addr
```

File: conn_native.go (hunk 3 above)
Comment (albrow) [reply conn_native.go]: That kind of just kicks the can down the road. How would we know whether the connection was secure (WSS) or insecure (WS) when constructing a `Conn` for [new incoming connections](https://github.com/libp2p/go-ws-transport/blob/ea33ee6f3fc5c4827c1a09cdf2ace5f705714fef/listener.go#L36)? Is it safe to always assume incoming connections are insecure, even if you were actually dialed at a WSS address via a proxy like nginx? Does this throw a wrench in our plan of using a proxy?

File: addrs.go (hunk 8 above)
Comment (Stebalien) [reply addrs.go]: https://github.com/multiformats/go-multiaddr-net/pull/63

File: websocket_native.go (hunk 5 above)
Comment (Stebalien) [reply websocket_native.go]: We have to perform the security upgrade at the moment. We don't currently have a way to use the TLS connection's security. We'd probably need to allocate a new code (or use `/tls/ws`) if we decide to switch unless we want to break all js-libp2p nodes.

File: addrs.go (hunk 8 above)
Comment (raulk) [reply addrs.go]: Merged and released: https://github.com/multiformats/go-multiaddr-net/releases/tag/v0.1.2. @albrow, mind updating and testing?

**Comment by raulk** [comment]
Oops, GitHub automation had one up their sleeve.

File: addrs.go (hunk 8 above)
Comment (albrow) [reply addrs.go]: Thanks. @raulk @Stebalien. Just updated and the tests pass.

**Comment by Stebalien** [comment]
PR to move wss protocol definition: https://github.com/0xProject/go-ws-transport/pull/2

File: addrs.go (hunk 1 above)
Comment (Stebalien) [reply addrs.go]: > Does libp2p need to resolve the domain name here for some reason?

No, and it shouldn't resolve it.

> Suppose a contains a domain name instead of an IP address. How can I differentiate between dns4 and dns6?

You can now use `/dns`.

Comment (Stebalien) [reply addrs.go]: Oh, resolved below.

File: conn_browser.go [hunk 11]
```diff
@@ -169,9 +170,32 @@ func (c *Conn) LocalAddr() net.Addr {
 
 func getRemoteAddr(val js.Value) net.Addr {
 	rawURL := val.Get("url").String()
+	if strings.HasPrefix(rawURL, "ws://") {
+		return getRemoteAddrWs(rawURL)
+	} else {
+		// Assume WSS
+		return getRemoteAddrWss(rawURL)
+	}
```
Comment (Stebalien) [inline conn_browser.go]: Can you not just parse this as a URL and create an `Addr{URL: ...}`?

File: websocket_native.go [hunk 12]
```diff
@@ -5,21 +5,53 @@ package websocket
...
+
+	// Apply functional options.
+	for _, o := range opts {
+		o(c)
+	}
+
+	// Configure ws.Dialer based on given TLSClientConfig
+	dialer := new(ws.Dialer)
+	(*dialer) = *ws.DefaultDialer
+	dialer.TLSClientConfig = c.TLSClientConfig
```
Comment (Stebalien) [inline websocket_native.go]: nit:

```go
dialer := *ws.DefaultDialer
```

?

(you can then use `&` to take a reference below).

File: conn_native.go (hunk 3 above)
Comment (Stebalien) [reply conn_native.go]: https://github.com/0xProject/go-ws-transport/pull/4

(could use some tests)

**Comment by Stebalien** [comment]
Something I realized when we were discussing using an nginx reverse proxy. `transport.Listen("/..../wss")` should return an error.

File: options_native.go [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	"crypto/tls"
+	"net/http"
+
+	"github.com/gorilla/websocket"
+)
+
+// WebsocketConfig is the configuration type for the Websocket transport.
+type WebsocketConfig struct {
```
Comment (marten-seemann) [inline options_native.go:13]: If this is an exported struct, we don't need the Option constructor (and vice versa).

File: websocket_browser.go [hunk 14]
```diff
@@ -5,26 +5,73 @@ package websocket
...
+			Upgrader: u,
+			Config:   c,
+		}
+		return t
+	}
+}
+
+func (t *WebsocketTransport) maDial(ctx context.Context, raddr ma.Multiaddr) (mnc manet.Conn, err error) {
+	defer func() {
+		if r := recover(); r != nil {
```
Comment (marten-seemann) [inline websocket_browser.go:52]: What is this used for?

File: addrs.go [hunk 15]
```diff
@@ -21,22 +53,36 @@ func (addr *Addr) Network() string {
...
+// Deprecated. Use NewAddrWithScheme.
 func NewAddr(host string) *Addr {
+	// Older versions of the transport only supported insecure connections (i.e.
+	// WS instead of WSS). Assume that is the case here.
+	return NewAddrWithScheme(host, false)
+}
+
+// NewAddrWithScheme creates a new Addr using the given host string. isSecure
+// should be true for WSS connections and false for WS.
+func NewAddrWithScheme(host string, isSecure bool) *Addr {
```
Comment (marten-seemann) [inline addrs.go:67]: Do we really need an `Addr` constructor? It seems quite trivial to pass in a `url.URL`.

**Review by marten-seemann — COMMENTED** [review]
First quick review.

File: websocket_browser.go (hunk 14 above)
Comment (Stebalien) [reply websocket_browser.go:52]: IIRC, we were concerned about some panics in the underlying library. I don't know if those have been fixed.
